ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs

```diff
@@ -16,59 +13,49 @@ import           Ouroboros.Consensus.Ledger.Abstract
 import           Ouroboros.Consensus.Ledger.Extended
 import           Ouroboros.Consensus.Util.IOLike
 
-import           Ouroboros.Consensus.Storage.ChainDB (LedgerCursor (..),
-                     LedgerCursorFailure (..))
-
-localStateQueryServer
-  :: forall m blk. (IOLike m, QueryLedger blk)
+localStateQueryServer ::
+     forall m blk. (IOLike m, QueryLedger blk)
   => LedgerConfig blk
-  -> m (LedgerCursor m blk)
+  -> (Point blk -> STM m (Maybe (ExtLedgerState blk)))
+     -- ^ Get a past ledger
+  -> STM m (Point blk)
+     -- ^ Get the immutable point
   -> LocalStateQueryServer blk (Query blk) m ()
-localStateQueryServer cfg newLedgerCursor =
-    LocalStateQueryServer $ idle <$> newLedgerCursor
+localStateQueryServer cfg getPastLedger getImmutablePoint =
+    LocalStateQueryServer $ return idle
   where
-    idle
-      :: LedgerCursor m blk
-      -> ServerStIdle blk (Query blk) m ()
-    idle ledgerCursor = ServerStIdle
-      { recvMsgAcquire = handleAcquire ledgerCursor
-      , recvMsgDone    = return ()
-      }
-
-    handleAcquire
-      :: LedgerCursor m blk
-      -> Point blk
-      -> m (ServerStAcquiring blk (Query blk) m ())
-    handleAcquire ledgerCursor pt =
-      ledgerCursorMove ledgerCursor pt <&> \case
-        Left failure ->
-          SendMsgFailure (translateFailure failure) (idle ledgerCursor)
-        Right ExtLedgerState { ledgerState } ->
-          SendMsgAcquired (acquired ledgerState ledgerCursor)
-
-    acquired
-      :: LedgerState blk
-      -> LedgerCursor m blk
-      -> ServerStAcquired blk (Query blk) m ()
-    acquired ledgerState ledgerCursor = ServerStAcquired
-      { recvMsgQuery     = handleQuery ledgerState ledgerCursor
-      , recvMsgReAcquire = handleAcquire ledgerCursor
-      , recvMsgRelease   = return $ idle ledgerCursor
-      }
-
-    handleQuery
-      :: LedgerState blk
-      -> LedgerCursor m blk
+    idle :: ServerStIdle blk (Query blk) m ()
+    idle = ServerStIdle {
+          recvMsgAcquire = handleAcquire
+        , recvMsgDone    = return ()
+        }
+
+    handleAcquire :: Point blk -> m (ServerStAcquiring blk (Query blk) m ())
+    handleAcquire pt = do
+        (mPastLedger, immutablePoint) <-
+          atomically $ (,) <$> getPastLedger pt <*> getImmutablePoint
+
+        return $ case mPastLedger of
+          Just pastLedger
+            -> SendMsgAcquired $ acquired (ledgerState pastLedger)
+          Nothing
+            | pointSlot pt < pointSlot immutablePoint
+            -> SendMsgFailure AcquireFailurePointTooOld idle
+            | otherwise
+            -> SendMsgFailure AcquireFailurePointNotOnChain idle
+
+    acquired :: LedgerState blk -> ServerStAcquired blk (Query blk) m ()
+    acquired ledgerState = ServerStAcquired {
+          recvMsgQuery     = handleQuery ledgerState
```

Comment:
An idle question: were these message names too tied to the notion of a cursor?

Reply:
If I remember correctly, it's actually the other way round. The cursor was written for the `LocalStateQuery` protocol. But the `LocalStateQuery` protocol was written in mind with the LedgerDB and its (previous) cost model in mind: acquiring a snapshot was not cheap (in terms of time, but cheap in terms of space, thanks to sharing), but is now.